lms/views/decorators/h_api.py

```diff
@@ -121,73 +121,54 @@ def wrapper(request, jwt, context=None):
             return wrapped(request, jwt)
 
         context = context or request.context
-        get_param = functools.partial(_get_param, request)
-
-        tool_consumer_instance_guid = get_param("tool_consumer_instance_guid")
-        context_id = get_param("context_id")
-
-        group = models.CourseGroup.get(
-            request.db, tool_consumer_instance_guid, context_id
-        )
-
-        assert group is not None, (
-            "create_course_group() should always have "
-            "been run successfully before this "
-            "function gets called, so group should "
-            "never be None."
-        )
 
         request.find_service(name="hapi").post(
-            f"groups/{group.pubid}/members/{context.h_userid}"
+            f"groups/{context.h_pubid}/members/{context.h_userid}"
         )
 
         return wrapped(request, jwt)
 
     return wrapper
 
 
-def _maybe_create_group(context, request):
-    """Create a Hypothesis group for the LTI course, if one doesn't exist."""
+def _upsert_group(context, request):
+    """
+    Create or update the course group in h.
+
+    This also sets the ``context.h_pubid`` attribute to the group's id received
+    from h.
+    """
     # Only create groups for application instances that we've enabled the
     # auto provisioning features for.
     if not _auto_provisioning_feature_enabled(request):
         return
 
     get_param = functools.partial(_get_param, request)
 
-    tool_consumer_instance_guid = get_param("tool_consumer_instance_guid")
-    context_id = get_param("context_id")
-
-    group = models.CourseGroup.get(request.db, tool_consumer_instance_guid, context_id)
-
-    if group:
-        # The group has already been created in h, nothing more to do here.
-        return
+    hapi_svc = request.find_service(name="hapi")
 
-    # Show the user an error page if the group hasn't been created yet and
-    # the user isn't allowed to create groups.
-    if not any(
+    is_instructor = any(
         role in get_param("roles").lower()
         for role in ("administrator", "instructor", "teachingassisstant")
-    ):
-        raise HTTPBadRequest("Instructor must launch assignment first.")
-
-    # Create the group in h.
-    response = request.find_service(name="hapi").post(
-        "groups",
-        {"groupid": context.h_groupid, "name": context.h_group_name},
-        context.h_userid,
     )
 
-    # Save a record of the group's pubid in the DB so that we can find it
-    # again later.
-    request.db.add(
-        models.CourseGroup(
-            tool_consumer_instance_guid=tool_consumer_instance_guid,
-            context_id=context_id,
-            pubid=response.json()["id"],
+    try:
+        response = hapi_svc.patch(
+            f"groups/{context.h_groupid}", {"name": context.h_group_name}
         )
-    )
+    except HAPINotFoundError:
+        # The group hasn't been created in h yet.
```

Comment:
Technically I think the h API might also 404 if the user wasn't authorized to update the group (a problem with the API that I know Lyza is working on). In practice the LMS app is authorized to update groups -- it has a valid client id and secret, and doesn't try to update groups outside of its own authority. So the only reason the LMS app would get a 404 is if no group with the given groupid exists yet because the app hasn't created it yet.

Reply:
This is correct, but not something I was planning on changing (your wording "a problem with the API that I know Lyza is working on" is throwing me here).